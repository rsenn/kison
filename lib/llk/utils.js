// @ts-check

const data = require('../data');

let { cachedStateMatchMap } = data;

function setStateCacheMatch(state, tokensLength) {
  return
  let holder = cachedStateMatchMap.get(state);
  if (!holder) {
    holder = new Set();
    cachedStateMatchMap.set(state, holder);
  }
  holder.add(tokensLength);
}

function getStateCacheMatch(state, tokensLength) {
  return;
  // consumed token length may be different!
  const holder = cachedStateMatchMap.get(state);
  return holder?.has(tokensLength);
}

function clearStateMatchCache() {
  return
  cachedStateMatchMap = new Map();
}

module.exports = {
  setStateCacheMatch,
  getStateCacheMatch,
  clearStateMatchCache,
};