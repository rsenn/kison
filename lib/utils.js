/**
 * @ignore
 * utils for kison.
 * @author yiminghe@gmail.com
 */

var util = require("modulex-util");

var doubleReg = /"/g;
var single = /'/g;

const gened = {
  mix(to, from) {
    for (var f in from) {
      to[f] = from[f];
    }
  },

  isArray(obj) {
    return "[object Array]" === Object.prototype.toString.call(obj);
  },

  inArray(item, arr) {
    for (var i = 0, l = arr.length; i < l; i++) {
      if (arr[i] === item) {
        return true;
      }
    }
    return false;
  },

  each(object, fn, context) {
    if (object) {
      var key,
        val,
        length,
        i = 0;

      context = context || null;

      if (!isArray(object)) {
        for (key in object) {
          // can not use hasOwnProperty
          if (fn.call(context, object[key], key, object) === false) {
            break;
          }
        }
      } else {
        length = object.length;
        for (val = object[0]; i < length; val = object[++i]) {
          if (fn.call(context, val, i, object) === false) {
            break;
          }
        }
      }
    }
  }
};

var { isArray } = gened;

const utils = (module.exports = Object.assign({}, util, gened, {
  eachRhs(rhs, fn) {
    utils.each(rhs, (r, index) => {
      if (typeof r !== "string") {
        return;
      }
      fn(r, index);
    });
  },

  assertSymbolString(symbol) {
    if (typeof symbol !== "string") {
      throw new Error("invalid symbol".symbol);
    }
  },

  filterRhs(rhs) {
    return rhs.filter(r => typeof r === "string");
  },

  toFunctionString(fn, name) {
    const str = fn.toString();
    if (str.match(/^function\s+\(/)) {
      return str.replace(/^function\s+\(/, `function ${name || fn.name}(`);
    }
    return "function " + str;
  },

  escapeString(str, quote) {
    var regexp = single;
    if (quote === '"') {
      regexp = doubleReg;
    } else {
      quote = "'";
    }
    return str
      .replace(/\\/g, "\\\\")
      .replace(/\r/g, "\\r")
      .replace(/\n/g, "\\n")
      .replace(/\t/g, "\\t")
      .replace(regexp, "\\" + quote);
  },

  serializeObject(obj, excludeReg) {
    var r;

    if (
      excludeReg &&
      typeof excludeReg === "function" &&
      (r = excludeReg(obj)) === false
    ) {
      return false;
    }

    if (r !== undefined) {
      obj = r;
    }

    var ret = [];

    if (typeof obj === "string") {
      return "'" + utils.escapeString(obj) + "'";
    } else if (typeof obj === "number") {
      return obj + "";
    } else if (util.isRegExp(obj)) {
      return (
        "/" +
        obj.source +
        "/" +
        (obj.global ? "g" : "") +
        (obj.ignoreCase ? "i" : "") +
        (obj.multiline ? "m" : "")
      );
    } else if (util.isArray(obj)) {
      ret.push("[");
      var sub = [];
      util.each(obj, function(v) {
        var t = utils.serializeObject(v, excludeReg);
        if (t !== false) {
          sub.push(t);
        }
      });
      ret.push(sub.join(", "));
      ret.push("]");
      return ret.join("");
    } else if (typeof obj === "object") {
      ret = [];
      ret[0] = "{";
      var start = 1;
      for (var i in obj) {
        var v = obj[i];
        if (excludeReg && util.isRegExp(excludeReg) && i.match(excludeReg)) {
          continue;
        }
        var t = utils.serializeObject(v, excludeReg);
        if (t === false) {
          continue;
        }
        /*jshint quotmark:false*/
        var key = "'" + utils.escapeString(i) + "'";
        ret.push((start ? "" : ",") + key + ": " + t);
        start = 0;
      }
      ret.push("}");
      return ret.join("\n");
    } else if (typeof obj === "function") {
      return (obj + "").replace(/^[\w-]+/, "function");
    } else {
      return obj + "";
    }
  }
}));
