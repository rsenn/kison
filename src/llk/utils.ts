// @ts-check

import data from '../data';

let { productionsBySymbol } = data;

function isSymbol(s: string) {
  return !!productionsBySymbol[s];
}

export { isSymbol };
