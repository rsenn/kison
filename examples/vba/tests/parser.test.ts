import { prettyJson } from '../../../__tests__/utils';
import parser from '../src/parserLLK';

describe('vba parser', () => {
  it('parse correctly', () => {
    const code = `
sub main
MsgBox 1
end sub
`;
    const ret = parser.parse(code);

    expect(ret.error?.errorMessage).toMatchInlineSnapshot(`undefined`);

    expect(prettyJson(ret.ast)).toMatchSnapshot();
  });

  it('parse comment and line continuation correctly', () => {
    const code = `
sub main
' x _
sdfasdf
rem xxxx _
yyy

x = _
1

end sub
`;
    const ret = parser.parse(code);

    expect(ret.error?.errorMessage).toMatchInlineSnapshot(`undefined`);

    expect(prettyJson(ret.ast)).toMatchSnapshot();
  });
});
