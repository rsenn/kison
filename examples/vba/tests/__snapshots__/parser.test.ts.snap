// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vba parser parse comment and line continuation correctly 2`] = `
"{
  'start': 0,
  'end': 57,
  'firstLine': 1,
  'lastLine': 12,
  'firstColumn': 1,
  'lastColumn': 1,
  'symbol': 'progam',
  'type': 'symbol',
  'children': [
    {
      'start': 0,
      'end': 1,
      'firstLine': 1,
      'lastLine': 2,
      'firstColumn': 1,
      'lastColumn': 1,
      'symbol': 'endOfLine',
      'type': 'symbol',
      'children': [
        {
          'start': 0,
          'end': 1,
          'firstLine': 1,
          'lastLine': 2,
          'firstColumn': 1,
          'lastColumn': 1,
          'token': 'NEWLINE',
          'type': 'token',
          'text': '\\\\n'
        }
      ],
      'ruleIndex': 2,
      'internalRuleIndex': 2,
      'id': 3
    },
    {
      'start': 1,
      'end': 57,
      'firstLine': 2,
      'lastLine': 12,
      'firstColumn': 1,
      'lastColumn': 1,
      'symbol': 'moduleBody',
      'type': 'symbol',
      'children': [
        {
          'start': 1,
          'end': 56,
          'firstLine': 2,
          'lastLine': 11,
          'firstColumn': 1,
          'lastColumn': 8,
          'symbol': 'moduleBodyElement',
          'type': 'symbol',
          'children': [
            {
              'start': 1,
              'end': 56,
              'firstLine': 2,
              'lastLine': 11,
              'firstColumn': 1,
              'lastColumn': 8,
              'symbol': 'subStmt',
              'type': 'symbol',
              'children': [
                {
                  'start': 1,
                  'end': 4,
                  'firstLine': 2,
                  'lastLine': 2,
                  'firstColumn': 1,
                  'lastColumn': 4,
                  'token': 'SUB',
                  'type': 'token',
                  'text': 'sub'
                },
                {
                  'start': 5,
                  'end': 9,
                  'firstLine': 2,
                  'lastLine': 2,
                  'firstColumn': 5,
                  'lastColumn': 9,
                  'symbol': 'ambiguousIdentifier',
                  'type': 'symbol',
                  'children': [
                    {
                      'start': 5,
                      'end': 9,
                      'firstLine': 2,
                      'lastLine': 2,
                      'firstColumn': 5,
                      'lastColumn': 9,
                      'token': 'IDENTIFIER',
                      'type': 'token',
                      'text': 'main'
                    }
                  ],
                  'ruleIndex': 55,
                  'internalRuleIndex': 162,
                  'id': 7
                },
                {
                  'start': 9,
                  'end': 40,
                  'firstLine': 2,
                  'lastLine': 8,
                  'firstColumn': 9,
                  'lastColumn': 1,
                  'symbol': 'endOfStatement',
                  'type': 'symbol',
                  'children': [
                    {
                      'start': 9,
                      'end': 10,
                      'firstLine': 2,
                      'lastLine': 3,
                      'firstColumn': 9,
                      'lastColumn': 1,
                      'symbol': 'endOfLine',
                      'type': 'symbol',
                      'children': [
                        {
                          'start': 9,
                          'end': 10,
                          'firstLine': 2,
                          'lastLine': 3,
                          'firstColumn': 9,
                          'lastColumn': 1,
                          'token': 'NEWLINE',
                          'type': 'token',
                          'text': '\\\\n'
                        }
                      ],
                      'ruleIndex': 2,
                      'internalRuleIndex': 2,
                      'id': 9
                    },
                    {
                      'start': 10,
                      'end': 23,
                      'firstLine': 3,
                      'lastLine': 4,
                      'firstColumn': 1,
                      'lastColumn': 8,
                      'symbol': 'endOfLine',
                      'type': 'symbol',
                      'children': [
                        {
                          'start': 10,
                          'end': 23,
                          'firstLine': 3,
                          'lastLine': 4,
                          'firstColumn': 1,
                          'lastColumn': 8,
                          'token': 'COMMENT',
                          'type': 'token',
                          'text': '' x _\\\\nsdfasdf'
                        }
                      ],
                      'ruleIndex': 2,
                      'internalRuleIndex': 3,
                      'id': 10
                    },
                    {
                      'start': 23,
                      'end': 24,
                      'firstLine': 4,
                      'lastLine': 5,
                      'firstColumn': 8,
                      'lastColumn': 1,
                      'symbol': 'endOfLine',
                      'type': 'symbol',
                      'children': [
                        {
                          'start': 23,
                          'end': 24,
                          'firstLine': 4,
                          'lastLine': 5,
                          'firstColumn': 8,
                          'lastColumn': 1,
                          'token': 'NEWLINE',
                          'type': 'token',
                          'text': '\\\\n'
                        }
                      ],
                      'ruleIndex': 2,
                      'internalRuleIndex': 2,
                      'id': 11
                    },
                    {
                      'start': 24,
                      'end': 38,
                      'firstLine': 5,
                      'lastLine': 6,
                      'firstColumn': 1,
                      'lastColumn': 4,
                      'symbol': 'endOfLine',
                      'type': 'symbol',
                      'children': [
                        {
                          'start': 24,
                          'end': 38,
                          'firstLine': 5,
                          'lastLine': 6,
                          'firstColumn': 1,
                          'lastColumn': 4,
                          'token': 'REMCOMMENT',
                          'type': 'token',
                          'text': 'rem xxxx _\\\\nyyy'
                        }
                      ],
                      'ruleIndex': 2,
                      'internalRuleIndex': 4,
                      'id': 12
                    },
                    {
                      'start': 38,
                      'end': 40,
                      'firstLine': 6,
                      'lastLine': 8,
                      'firstColumn': 4,
                      'lastColumn': 1,
                      'symbol': 'endOfLine',
                      'type': 'symbol',
                      'children': [
                        {
                          'start': 38,
                          'end': 40,
                          'firstLine': 6,
                          'lastLine': 8,
                          'firstColumn': 4,
                          'lastColumn': 1,
                          'token': 'NEWLINE',
                          'type': 'token',
                          'text': '\\\\n\\\\n'
                        }
                      ],
                      'ruleIndex': 2,
                      'internalRuleIndex': 2,
                      'id': 13
                    }
                  ],
                  'ruleIndex': 3,
                  'internalRuleIndex': 9,
                  'id': 8
                },
                {
                  'start': 40,
                  'end': 49,
                  'firstLine': 8,
                  'lastLine': 11,
                  'firstColumn': 1,
                  'lastColumn': 1,
                  'symbol': 'block',
                  'type': 'symbol',
                  'children': [
                    {
                      'start': 40,
                      'end': 47,
                      'firstLine': 8,
                      'lastLine': 9,
                      'firstColumn': 1,
                      'lastColumn': 2,
                      'symbol': 'blockStmt',
                      'type': 'symbol',
                      'children': [
                        {
                          'start': 40,
                          'end': 47,
                          'firstLine': 8,
                          'lastLine': 9,
                          'firstColumn': 1,
                          'lastColumn': 2,
                          'symbol': 'letStmt',
                          'type': 'symbol',
                          'children': [
                            {
                              'start': 40,
                              'end': 41,
                              'firstLine': 8,
                              'lastLine': 8,
                              'firstColumn': 1,
                              'lastColumn': 2,
                              'symbol': 'implicitCallStmt_InStmt',
                              'type': 'symbol',
                              'children': [
                                {
                                  'start': 40,
                                  'end': 41,
                                  'firstLine': 8,
                                  'lastLine': 8,
                                  'firstColumn': 1,
                                  'lastColumn': 2,
                                  'symbol': 'iCS_S_VariableOrProcedureCall',
                                  'type': 'symbol',
                                  'children': [
                                    {
                                      'start': 40,
                                      'end': 41,
                                      'firstLine': 8,
                                      'lastLine': 8,
                                      'firstColumn': 1,
                                      'lastColumn': 2,
                                      'symbol': 'ambiguousIdentifier',
                                      'type': 'symbol',
                                      'children': [
                                        {
                                          'start': 40,
                                          'end': 41,
                                          'firstLine': 8,
                                          'lastLine': 8,
                                          'firstColumn': 1,
                                          'lastColumn': 2,
                                          'token': 'IDENTIFIER',
                                          'type': 'token',
                                          'text': 'x'
                                        }
                                      ],
                                      'ruleIndex': 55,
                                      'internalRuleIndex': 162,
                                      'id': 19
                                    }
                                  ],
                                  'ruleIndex': 44,
                                  'internalRuleIndex': 122,
                                  'id': 18
                                }
                              ],
                              'ruleIndex': 40,
                              'internalRuleIndex': 107,
                              'id': 17
                            },
                            {
                              'start': 42,
                              'end': 43,
                              'firstLine': 8,
                              'lastLine': 8,
                              'firstColumn': 3,
                              'lastColumn': 4,
                              'token': 'EQ',
                              'type': 'token',
                              'text': '='
                            },
                            {
                              'start': 46,
                              'end': 47,
                              'firstLine': 9,
                              'lastLine': 9,
                              'firstColumn': 1,
                              'lastColumn': 2,
                              'symbol': 'valueStmt',
                              'type': 'symbol',
                              'children': [
                                {
                                  'start': 46,
                                  'end': 47,
                                  'firstLine': 9,
                                  'lastLine': 9,
                                  'firstColumn': 1,
                                  'lastColumn': 2,
                                  'symbol': 'literal',
                                  'type': 'symbol',
                                  'children': [
                                    {
                                      'start': 46,
                                      'end': 47,
                                      'firstLine': 9,
                                      'lastLine': 9,
                                      'firstColumn': 1,
                                      'lastColumn': 2,
                                      'token': 'INTEGERLITERAL',
                                      'type': 'token',
                                      'text': '1'
                                    }
                                  ],
                                  'ruleIndex': 46,
                                  'internalRuleIndex': 124,
                                  'id': 21
                                }
                              ],
                              'ruleIndex': 39,
                              'internalRuleIndex': 103,
                              'id': 20
                            }
                          ],
                          'ruleIndex': 21,
                          'internalRuleIndex': 62,
                          'id': 16
                        }
                      ],
                      'ruleIndex': 16,
                      'internalRuleIndex': 42,
                      'id': 15
                    },
                    {
                      'start': 47,
                      'end': 49,
                      'firstLine': 9,
                      'lastLine': 11,
                      'firstColumn': 2,
                      'lastColumn': 1,
                      'symbol': 'endOfStatement',
                      'type': 'symbol',
                      'children': [
                        {
                          'start': 47,
                          'end': 49,
                          'firstLine': 9,
                          'lastLine': 11,
                          'firstColumn': 2,
                          'lastColumn': 1,
                          'symbol': 'endOfLine',
                          'type': 'symbol',
                          'children': [
                            {
                              'start': 47,
                              'end': 49,
                              'firstLine': 9,
                              'lastLine': 11,
                              'firstColumn': 2,
                              'lastColumn': 1,
                              'token': 'NEWLINE',
                              'type': 'token',
                              'text': '\\\\n\\\\n'
                            }
                          ],
                          'ruleIndex': 2,
                          'internalRuleIndex': 2,
                          'id': 23
                        }
                      ],
                      'ruleIndex': 3,
                      'internalRuleIndex': 9,
                      'id': 22
                    }
                  ],
                  'ruleIndex': 15,
                  'internalRuleIndex': 36,
                  'id': 14
                },
                {
                  'start': 49,
                  'end': 56,
                  'firstLine': 11,
                  'lastLine': 11,
                  'firstColumn': 1,
                  'lastColumn': 8,
                  'token': 'END_SUB',
                  'type': 'token',
                  'text': 'end sub'
                }
              ],
              'ruleIndex': 10,
              'internalRuleIndex': 30,
              'id': 6
            }
          ],
          'ruleIndex': 8,
          'internalRuleIndex': 25,
          'id': 5
        },
        {
          'start': 56,
          'end': 57,
          'firstLine': 11,
          'lastLine': 12,
          'firstColumn': 8,
          'lastColumn': 1,
          'symbol': 'endOfLine',
          'type': 'symbol',
          'children': [
            {
              'start': 56,
              'end': 57,
              'firstLine': 11,
              'lastLine': 12,
              'firstColumn': 8,
              'lastColumn': 1,
              'token': 'NEWLINE',
              'type': 'token',
              'text': '\\\\n'
            }
          ],
          'ruleIndex': 2,
          'internalRuleIndex': 2,
          'id': 24
        }
      ],
      'ruleIndex': 7,
      'internalRuleIndex': 20,
      'id': 4
    },
    {
      'start': 57,
      'end': 57,
      'firstLine': 12,
      'lastLine': 12,
      'firstColumn': 1,
      'lastColumn': 1,
      'token': '$EOF',
      'type': 'token',
      'text': ''
    }
  ],
  'ruleIndex': 1,
  'internalRuleIndex': 1,
  'id': 2
}"
`;

exports[`vba parser parse correctly 2`] = `
"{
  'start': 0,
  'end': 27,
  'firstLine': 1,
  'lastLine': 5,
  'firstColumn': 1,
  'lastColumn': 1,
  'symbol': 'progam',
  'type': 'symbol',
  'children': [
    {
      'start': 0,
      'end': 1,
      'firstLine': 1,
      'lastLine': 2,
      'firstColumn': 1,
      'lastColumn': 1,
      'symbol': 'endOfLine',
      'type': 'symbol',
      'children': [
        {
          'start': 0,
          'end': 1,
          'firstLine': 1,
          'lastLine': 2,
          'firstColumn': 1,
          'lastColumn': 1,
          'token': 'NEWLINE',
          'type': 'token',
          'text': '\\\\n'
        }
      ],
      'ruleIndex': 2,
      'internalRuleIndex': 2,
      'id': 3
    },
    {
      'start': 1,
      'end': 27,
      'firstLine': 2,
      'lastLine': 5,
      'firstColumn': 1,
      'lastColumn': 1,
      'symbol': 'moduleBody',
      'type': 'symbol',
      'children': [
        {
          'start': 1,
          'end': 26,
          'firstLine': 2,
          'lastLine': 4,
          'firstColumn': 1,
          'lastColumn': 8,
          'symbol': 'moduleBodyElement',
          'type': 'symbol',
          'children': [
            {
              'start': 1,
              'end': 26,
              'firstLine': 2,
              'lastLine': 4,
              'firstColumn': 1,
              'lastColumn': 8,
              'symbol': 'subStmt',
              'type': 'symbol',
              'children': [
                {
                  'start': 1,
                  'end': 4,
                  'firstLine': 2,
                  'lastLine': 2,
                  'firstColumn': 1,
                  'lastColumn': 4,
                  'token': 'SUB',
                  'type': 'token',
                  'text': 'sub'
                },
                {
                  'start': 5,
                  'end': 9,
                  'firstLine': 2,
                  'lastLine': 2,
                  'firstColumn': 5,
                  'lastColumn': 9,
                  'symbol': 'ambiguousIdentifier',
                  'type': 'symbol',
                  'children': [
                    {
                      'start': 5,
                      'end': 9,
                      'firstLine': 2,
                      'lastLine': 2,
                      'firstColumn': 5,
                      'lastColumn': 9,
                      'token': 'IDENTIFIER',
                      'type': 'token',
                      'text': 'main'
                    }
                  ],
                  'ruleIndex': 55,
                  'internalRuleIndex': 162,
                  'id': 7
                },
                {
                  'start': 9,
                  'end': 10,
                  'firstLine': 2,
                  'lastLine': 3,
                  'firstColumn': 9,
                  'lastColumn': 1,
                  'symbol': 'endOfStatement',
                  'type': 'symbol',
                  'children': [
                    {
                      'start': 9,
                      'end': 10,
                      'firstLine': 2,
                      'lastLine': 3,
                      'firstColumn': 9,
                      'lastColumn': 1,
                      'symbol': 'endOfLine',
                      'type': 'symbol',
                      'children': [
                        {
                          'start': 9,
                          'end': 10,
                          'firstLine': 2,
                          'lastLine': 3,
                          'firstColumn': 9,
                          'lastColumn': 1,
                          'token': 'NEWLINE',
                          'type': 'token',
                          'text': '\\\\n'
                        }
                      ],
                      'ruleIndex': 2,
                      'internalRuleIndex': 2,
                      'id': 9
                    }
                  ],
                  'ruleIndex': 3,
                  'internalRuleIndex': 9,
                  'id': 8
                },
                {
                  'start': 10,
                  'end': 19,
                  'firstLine': 3,
                  'lastLine': 4,
                  'firstColumn': 1,
                  'lastColumn': 1,
                  'symbol': 'block',
                  'type': 'symbol',
                  'children': [
                    {
                      'start': 10,
                      'end': 18,
                      'firstLine': 3,
                      'lastLine': 3,
                      'firstColumn': 1,
                      'lastColumn': 9,
                      'symbol': 'blockStmt',
                      'type': 'symbol',
                      'children': [
                        {
                          'start': 10,
                          'end': 18,
                          'firstLine': 3,
                          'lastLine': 3,
                          'firstColumn': 1,
                          'lastColumn': 9,
                          'symbol': 'implicitCallStmt_InBlock',
                          'type': 'symbol',
                          'children': [
                            {
                              'start': 10,
                              'end': 18,
                              'firstLine': 3,
                              'lastLine': 3,
                              'firstColumn': 1,
                              'lastColumn': 9,
                              'symbol': 'iCS_B_ProcedureCall',
                              'type': 'symbol',
                              'children': [
                                {
                                  'start': 10,
                                  'end': 16,
                                  'firstLine': 3,
                                  'lastLine': 3,
                                  'firstColumn': 1,
                                  'lastColumn': 7,
                                  'symbol': 'ambiguousIdentifier',
                                  'type': 'symbol',
                                  'children': [
                                    {
                                      'start': 10,
                                      'end': 16,
                                      'firstLine': 3,
                                      'lastLine': 3,
                                      'firstColumn': 1,
                                      'lastColumn': 7,
                                      'token': 'IDENTIFIER',
                                      'type': 'token',
                                      'text': 'msgbox'
                                    }
                                  ],
                                  'ruleIndex': 55,
                                  'internalRuleIndex': 162,
                                  'id': 14
                                },
                                {
                                  'start': 17,
                                  'end': 18,
                                  'firstLine': 3,
                                  'lastLine': 3,
                                  'firstColumn': 8,
                                  'lastColumn': 9,
                                  'symbol': 'argsCall',
                                  'type': 'symbol',
                                  'children': [
                                    {
                                      'start': 17,
                                      'end': 18,
                                      'firstLine': 3,
                                      'lastLine': 3,
                                      'firstColumn': 8,
                                      'lastColumn': 9,
                                      'symbol': 'argCall',
                                      'type': 'symbol',
                                      'children': [
                                        {
                                          'start': 17,
                                          'end': 18,
                                          'firstLine': 3,
                                          'lastLine': 3,
                                          'firstColumn': 8,
                                          'lastColumn': 9,
                                          'symbol': 'valueStmt',
                                          'type': 'symbol',
                                          'children': [
                                            {
                                              'start': 17,
                                              'end': 18,
                                              'firstLine': 3,
                                              'lastLine': 3,
                                              'firstColumn': 8,
                                              'lastColumn': 9,
                                              'symbol': 'literal',
                                              'type': 'symbol',
                                              'children': [
                                                {
                                                  'start': 17,
                                                  'end': 18,
                                                  'firstLine': 3,
                                                  'lastLine': 3,
                                                  'firstColumn': 8,
                                                  'lastColumn': 9,
                                                  'token': 'INTEGERLITERAL',
                                                  'type': 'token',
                                                  'text': '1'
                                                }
                                              ],
                                              'ruleIndex': 46,
                                              'internalRuleIndex': 124,
                                              'id': 18
                                            }
                                          ],
                                          'ruleIndex': 39,
                                          'internalRuleIndex': 103,
                                          'id': 17
                                        }
                                      ],
                                      'ruleIndex': 30,
                                      'internalRuleIndex': 82,
                                      'id': 16
                                    }
                                  ],
                                  'ruleIndex': 29,
                                  'internalRuleIndex': 75,
                                  'id': 15
                                }
                              ],
                              'ruleIndex': 28,
                              'internalRuleIndex': 73,
                              'id': 13
                            }
                          ],
                          'ruleIndex': 26,
                          'internalRuleIndex': 71,
                          'id': 12
                        }
                      ],
                      'ruleIndex': 16,
                      'internalRuleIndex': 44,
                      'id': 11
                    },
                    {
                      'start': 18,
                      'end': 19,
                      'firstLine': 3,
                      'lastLine': 4,
                      'firstColumn': 9,
                      'lastColumn': 1,
                      'symbol': 'endOfStatement',
                      'type': 'symbol',
                      'children': [
                        {
                          'start': 18,
                          'end': 19,
                          'firstLine': 3,
                          'lastLine': 4,
                          'firstColumn': 9,
                          'lastColumn': 1,
                          'symbol': 'endOfLine',
                          'type': 'symbol',
                          'children': [
                            {
                              'start': 18,
                              'end': 19,
                              'firstLine': 3,
                              'lastLine': 4,
                              'firstColumn': 9,
                              'lastColumn': 1,
                              'token': 'NEWLINE',
                              'type': 'token',
                              'text': '\\\\n'
                            }
                          ],
                          'ruleIndex': 2,
                          'internalRuleIndex': 2,
                          'id': 20
                        }
                      ],
                      'ruleIndex': 3,
                      'internalRuleIndex': 9,
                      'id': 19
                    }
                  ],
                  'ruleIndex': 15,
                  'internalRuleIndex': 36,
                  'id': 10
                },
                {
                  'start': 19,
                  'end': 26,
                  'firstLine': 4,
                  'lastLine': 4,
                  'firstColumn': 1,
                  'lastColumn': 8,
                  'token': 'END_SUB',
                  'type': 'token',
                  'text': 'end sub'
                }
              ],
              'ruleIndex': 10,
              'internalRuleIndex': 30,
              'id': 6
            }
          ],
          'ruleIndex': 8,
          'internalRuleIndex': 25,
          'id': 5
        },
        {
          'start': 26,
          'end': 27,
          'firstLine': 4,
          'lastLine': 5,
          'firstColumn': 8,
          'lastColumn': 1,
          'symbol': 'endOfLine',
          'type': 'symbol',
          'children': [
            {
              'start': 26,
              'end': 27,
              'firstLine': 4,
              'lastLine': 5,
              'firstColumn': 8,
              'lastColumn': 1,
              'token': 'NEWLINE',
              'type': 'token',
              'text': '\\\\n'
            }
          ],
          'ruleIndex': 2,
          'internalRuleIndex': 2,
          'id': 21
        }
      ],
      'ruleIndex': 7,
      'internalRuleIndex': 20,
      'id': 4
    },
    {
      'start': 27,
      'end': 27,
      'firstLine': 5,
      'lastLine': 5,
      'firstColumn': 1,
      'lastColumn': 1,
      'token': '$EOF',
      'type': 'token',
      'text': ''
    }
  ],
  'ruleIndex': 1,
  'internalRuleIndex': 1,
  'id': 2
}"
`;
